packages/edit-site/src/components/sidebar-edit-mode/template-panel/hooks.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { useMemo } from '@wordpress/element';
+import { store as coreStore } from '@wordpress/core-data';
+import { parse } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { store as editSiteStore } from '../../../store';
+import { PATTERN_CORE_SOURCES, PATTERN_TYPES } from '../../../utils/constants';
+import { unlock } from '../../../lock-unlock';
+
+function injectThemeAttributeInBlockTemplateContent(
+	block,
+	currentThemeStylesheet
+) {
+	block.innerBlocks = block.innerBlocks.map( ( innerBlock ) => {
+		return injectThemeAttributeInBlockTemplateContent(
+			innerBlock,
+			currentThemeStylesheet
+		);
+	} );
+
+	if (
+		block.name === 'core/template-part' &&
+		block.attributes.theme === undefined
+	) {
+		block.attributes.theme = currentThemeStylesheet;
+	}
+	return block;
+}
+
+function preparePatterns( patterns, template, currentThemeStylesheet ) {
+	// Filter out duplicates.
+	const filterOutDuplicatesByName = ( currentItem, index, items ) =>
+		index === items.findIndex( ( item ) => currentItem.name === item.name );
+
+	// Filter out core patterns.
+	const filterOutCorePatterns = ( pattern ) =>
+		! PATTERN_CORE_SOURCES.includes( pattern.source );
+
+	// Filter only the patterns that are compatible with the current template.
+	const filterCompatiblePatterns = ( pattern ) =>
+		pattern.templateTypes?.includes( template.slug );
+
+	return patterns
+		.filter(
+			filterOutCorePatterns &&
+				filterOutDuplicatesByName &&
+				filterCompatiblePatterns
+		)
+		.map( ( pattern ) => ( {
+			...pattern,
+			keywords: pattern.keywords || [],
+			type: PATTERN_TYPES.theme,
+			blocks: parse( pattern.content, {
+				__unstableSkipMigrationLogs: true,
+			} ).map( ( block ) =>
+				injectThemeAttributeInBlockTemplateContent(
+					block,
+					currentThemeStylesheet
+				)
+			),
+		} ) );
+}
+
+export function useAvailablePatterns( template ) {
+	const { blockPatterns, restBlockPatterns, currentThemeStylesheet } =
+		useSelect( ( select ) => {
+			const { getSettings } = unlock( select( editSiteStore ) );
+			const settings = getSettings();
+
+			return {
+				blockPatterns:
+					settings.__experimentalAdditionalBlockPatterns ??
+					settings.__experimentalBlockPatterns,
+				restBlockPatterns: select( coreStore ).getBlockPatterns(),
+				currentThemeStylesheet:
+					select( coreStore ).getCurrentTheme().stylesheet,
+			};
+		}, [] );
+
+	return useMemo( () => {
+		const mergedPatterns = [
+			...( blockPatterns || [] ),
+			...( restBlockPatterns || [] ),
+		];
+		return preparePatterns(
+			mergedPatterns,
+			template,
+			currentThemeStylesheet
+		);
+	}, [ blockPatterns, restBlockPatterns, template, currentThemeStylesheet ] );
+}
```

Comment:
This file is a bit confusing to me. It's unfortunate that we have to do all this work in order to get the available list of patterns for a given template.

 - Isn't there a canonical way to retrieve block patterns?
 - Do we really need the "inject" stuff?

We need to have more clarity in terms of APIs there. How should a developer know that he has to do all of this in order to make use of patterns?

cc @ellatrix @jsnajdr as you both worked on patterns before as well.

Reply:
There is the `getAllPatterns` selector in `core/block-editor` store that could replace most of this `useAvailablePatterns` hook. But it has one major difference: it also loads all the reusable blocks, aka user patterns, using the `getReusableBlocks` selector. But the Site Editor is apparently never interested in user patterns. Does it even make sense to use user patterns in Site Editor? Or are they intended only for the Post Editor.

If we have a version of `getAllPatterns` that doesn't return user patterns, we could use it at all four places in Site Editor where the code that gathers patterns from various locations is repeated.

Reply:
> Do we really need the "inject" stuff?

Seems to me it was needed at the time when this PR was merged -- it was adding a `theme` attribute to all `core/template-part` blocks that didn't have a `theme` attribute. But then, one month later, @gziolo merged #55965 which added the `theme` fallback directly in the `core/template-part` block.

Today, I believe `injectThemeAttributeInBlockTemplateContent` could be removed. From more than one place.

Reply:
We would have to perform some extensive testing, but hopefully, injecting a `theme` attribute on the template level should no longer be necessary.